src/riot-rs-threads/src/scheduler.rs

```diff
@@ -0,0 +1,58 @@
+use critical_section::CriticalSection;
+
+use crate::THREADS;
+use crate::{Arch, Cpu};
+
+/// Schedule the next thread.
+///
+/// It selects the next thread that should run from the runqueue.
+/// This may be current thread, or a new one.
+///
+/// Input:
+/// - old_sp: the stack pointer of the currently running thread.
+///
+/// Returns:
+/// - `0` if the next thread in the runqueue is the currently running thread
+/// - Else: the next sp, and pointers to the old and new thread data are written into arch-specific registers
+#[no_mangle]
+pub unsafe fn sched(old_sp: usize) -> usize {
+    unsafe {
+        let cs = CriticalSection::new();
+        let next_pid;
+
+        loop {
+            {
+                if let Some(pid) = (&*THREADS.as_ptr(cs)).runqueue.get_next() {
+                    next_pid = pid;
+                    break;
+                }
+            }
+            Cpu::wfi();
+        }
+
+        let threads = &mut *THREADS.as_ptr(cs);
+        let current_high_regs;
+
+        if let Some(current_pid) = threads.current_pid() {
+            if next_pid == current_pid {
+                return 0;
+            }
+            //println!("current: {} next: {}", current_pid, next_pid);
+            threads.threads[current_pid as usize].sp = old_sp;
+            threads.current_thread = Some(next_pid);
+            current_high_regs = Some(&threads.threads[current_pid as usize].data);
+        } else {
+            current_high_regs = None;
+        }
+
+        let next = &threads.threads[next_pid as usize];
+        let next_sp = next.sp;
+        let next_high_regs = &next.data;
+
+        // Return pointers to the old and new thread data in arch specific registers
+        // so that the exception handler save and restore the state.
+        Cpu::return_data_in_regs(current_high_regs, next_high_regs);
```

Comment:
So this is (as it was and as it is now) obviously hacky as hell.
On the Cortex-M, it is possible to return two or four usize encoded in one u64 or one u128. [gotbolt](https://rust.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:2,endLineNumber:21,positionColumn:2,positionLineNumber:21,selectionStartColumn:2,selectionStartLineNumber:21,startColumn:2,startLineNumber:21),source:'%23!!%5Bno_std%5D%0A%0Apub+fn+fourints_tuple()+-%3E+(u32,+u32,+u32,+u32)+%7B%0A++++(0,1,2,3)%0A%7D%0A%0Apub+fn+fourints_u128()+-%3E+u128+%7B%0A++++3+%3C%3C+96+%7C+2+%3C%3C+64+%7C+1+%3C%3C+32+%7C+0%0A%7D%0A%0Apub+fn+twoints_u64()+-%3E+u64+%7B%0A++++2+%3C%3C+32+%7C+1%0A%7D%0A%0Apub+fn+threeints_u64u32()+-%3E+(u64,+u32)+%7B%0A++++(2+%3C%3C+32+%7C+1,+0)%0A%7D%0A%0Apub+fn+threeints_u32()+-%3E+(u32,+u32,+u32)+%7B%0A++++(2,+1,+0)%0A%7D'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1760,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'--target+thumbv7m-none-eabi',overrides:!((name:edition,value:'2021')),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+rustc+1.76.0+(Editor+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'rustc+1.66.0',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+rustc+1.76.0+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)
Unfortunately, it looks different on riscv32.

Reply:
With this PR I was trying to keep all logic/ implementation details the way that they currently are and just add some abstractions.
But I agree that it is hacky, and this new method definitely doesn't help with that.

Reply:
> With this PR I was trying to keep all logic/ implementation details the way that they currently are and just add some abstractions.

Which is really great, that way we see the impact of this abstraction (almost none).

Reply:
Marking this as resolved since I reverted the change that introduced `return_data_in_regs` (see https://github.com/future-proof-iot/RIOT-rs/pull/159#discussion_r1517414361).